proxy/src/connection.rs

```diff
@@ -254,6 +250,39 @@ impl Future for ConditionallyUpgradeServerToTls {
     }
 }
 
+impl ConditionallyUpgradeServerToTlsInner {
+    /// Polls the underlying socket for more data and buffers it.
+    ///
+    /// The buffer is matched for a TLS client hello message.
+    ///
+    /// `None` is returned if the underlying socket has closed.
+    fn poll_match_client_hello(&mut self) -> Poll<Option<tls::conditional_accept::Match>, io::Error> {
+        let sz = try_ready!(self.socket.read_buf(&mut self.peek_buf));
+        if sz == 0 {
+            return Ok(None.into())
+        }
+
+        let buf = self.peek_buf.as_ref();
+        Ok(Some(tls::conditional_accept::match_client_hello(buf, &self.tls.identity)).into())
+    }
+}
+
+impl Into<tls::UpgradeServerToTls> for ConditionallyUpgradeServerToTlsInner {
```

Comment:
Not a blocker, but to explain a comment you left earlier: if you switch these to `impl From<Conditionally...> for Thing`, then you get both `from` and `into` conversions.

Reply:
ahah!